src/Microsoft.ApiDesignGuidelines.Analyzers/Core/IdentifiersShouldNotHaveIncorrectSuffix.cs

```diff
@@ -56,13 +61,182 @@ public abstract class IdentifiersShouldNotHaveIncorrectSuffixAnalyzer : Diagnost
 
+        internal const string AttributeSuffix = "Attribute";
+        internal const string CollectionSuffix = "Collection";
+        internal const string DictionarySuffix = "Dictionary";
+        internal const string EventArgsSuffix = "EventArgs";
+        internal const string EventHandlerSuffix = "EventHandler";
+        internal const string ExSuffix = "Ex";
+        internal const string ExceptionSuffix = "Exception";
+        internal const string NewSuffix = "New";
+        internal const string PermissionSuffix = "Permission";
+        internal const string StreamSuffix = "Stream";
+        internal const string DelegateSuffix = "Delegate";
+        internal const string EnumSuffix = "Enum";
+        internal const string ImplSuffix = "Impl";
+        internal const string CoreSuffix = "Core";
+
+        // Dictionary from type name suffix to the type the suffix is required to denote.
+        private static readonly Dictionary<string, string> s_suffixToTypeDictionary = new Dictionary<string, string>
+        {
+            [AttributeSuffix] = "System.Attribute",
+            [CollectionSuffix] = "System.Collections.IEnumerable",
+            [DictionarySuffix] = "System.Collections.IDictionary",
+            [EventArgsSuffix] = "System.EventArgs",
+            [ExceptionSuffix] = "System.Exception",
+            [PermissionSuffix] = "System.Security.IPermission",
+            [StreamSuffix] = "System.IO.Stream"
+        };
+
+        // Dictionary from type name suffix to an array containing the only types that are
+        // allowed to have that suffix.
+        private static readonly Dictionary<string, string[]> s_suffixToAllowedTypesDictionary = new Dictionary<string, string[]>
+        {
+            [DelegateSuffix] = new[] { "System.Delegate", "System.MulticastDelegate" },
+            [EventHandlerSuffix] = new[] { "System.EventHandler" },
+            [EnumSuffix] = new[] { "System.Enum" }
+        };
+
         public override void Initialize(AnalysisContext analysisContext)
         {
+            analysisContext.RegisterSymbolAction(
+                (SymbolAnalysisContext context) =>
+                {
+                    var namedTypeSymbol = (INamedTypeSymbol)context.Symbol;
+                    if (namedTypeSymbol.GetResultantVisibility()!= SymbolVisibility.Public)
+                    {
+                        return;
+                    }
+
+                    string name = namedTypeSymbol.Name;
+                    Compilation compilation = context.Compilation;
+
+                    foreach (string suffix in s_suffixToTypeDictionary.Keys)
+                    {
+                        if (IsNotChildButHasSuffix(
+                            namedTypeSymbol,
+                            compilation.GetTypeByMetadataName(s_suffixToTypeDictionary[suffix]),
```

Comment:
We don't want to invoke `GetTypeByMetadataName` for each symbol. You should create a map upfront in CompilationStartAction, and access it in a nested Symbol action.